### ISSUE TITLE
Add native support for IIS appcmd

### ISSUE THREAD

**Issue by dagwieers** [ansible/ansible#25469]
_From @javadevmtl on June 7, 2016 21:10_

##### ISSUE TYPE

- Feature Idea
##### COMPONENT NAME

New module
##### ANSIBLE VERSION

```
Latest
```

Hi the default IIS module are ok for doing simple tasks. But the moment you need to configure anything complex you need to dive into using appcmd command.

For instance I have the following playbook command that registers an ISAPi DLL with IIS. The command works fine, but on secondary run because the task doesn't recognize the semantics of the command line it returns a failure. Basically it complains of a duplicate.

The command runs good on the first try:

```
    - name: Allow ISAPI DLL
      ignore_errors: yes
      raw: c:/Windows/System32/inetsrv/appcmd.exe set config -section:system.webServer/security/isapiCgiRestriction /+"[path='path-to-dll.dll',allowed='True',groupId='SomeGroupId',description='My DLL']" /commit:apphost
```

On second run:

```
[ERROR ( message:New add object missing required attributes. Cannot add duplicate collection entry of type 'add' with unique key attribute 'path' set to 'path-to-dll.dll',,. )]
```

We can ignore the error, but it would be nice if it could integrate natively with Ansible states such as "present", etc...

Here is some more info on appcmd: http://www.iis.net/learn/get-started/getting-started-with-iis/getting-started-with-appcmdexe

Basically it's a command line interface that lets you configure any part of IIS, which lends itself well for Ansible :)

_Copied from original issue: ansible/ansible-modules-extras#2386_

**Comment by dagwieers** [comment]
_From @ansibot on December 6, 2016 22:0_

This repository has been locked. All new issues and pullrequests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pullrequest to the ansible/ansible repo.

**Comment by jhawkesworth** [comment]
The IIS modules have improved significantly since this issue was raised.

We also now have win_command and win_shell which support creates, so you can do a limited sort of idempotency by using creates on your appcmd.exe invocations, and writing a file or creating a dir once all the appcmd.exe calls have completed the first time through.  Subsequent runs would then be skipped.

Also I believe IIS 10 has a rest configuration interface - see https://blogs.iis.net/adminapi/introducing-the-iis-administration-api and  https://github.com/Microsoft/IIS.Administration which presumably could be used via the uri / win_uri modules.

Yet another option is to make use of DSC resources - there seem to be many related to managing IIS here: http://www.powershellgallery.com/items?q=IIS+administration&x=0&y=0

**Comment by andrewsaraceni** [comment]
Overall, as @jhawkesworth noted, the current assortment of IIS modules handles a wide range of IIS-related items and tasks (sites, bindings, app pools, apps and virtual directories), and thus covers much of what could be done in appcmd.  The only missing piece I've noticed, which perhaps the original submitter alluded to, is a module for handling manipulation of the `applicationHost.config` file for global settings, such as for adding modules, setting anonymous authentication, etc.

Having something like a `win_iis_configuration` module that provides this functionality would be nice, which could be accomplished in a number of ways, the most cross-compatible I can think of being working with an instance of the `Microsoft.Web.Administration.ServerManager` class.

**Comment by jhawkesworth** [comment]
Just reviewing old issues and I think wrapping appcmd is not something we should want to do.
The original example illustrates how getting the command line syntax for it correct is at best tedious.

I think if there's still a specific need for an `win_iss_configuration` module as suggested by andrewsaraceni then we should have an open feature request for that, although if there is a suitable DSC resource - perhaps https://github.com/PowerShell/xWebAdministration/pull/330  that would probably be preferable to reinventing the feature here.

**Comment by jborean93** [comment]
Closing as per @jhawkesworth comments.
